mne/realtime/tests/test_fieldtrip_client.py

```diff
@@ -22,33 +24,95 @@
 
+    # sleep to make sure everything is setup before playback starts
     process = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE)
     # Let measurement continue for the entire duration
-    kill_signal.get(timeout=10.0)
+    kill_signal.get(timeout=40.0)
     process.terminate()
     with warnings.catch_warnings(record=True):  # still running
         process.stderr.close()
         process.stdout.close()
         del process
 
 
+def _start_buffer_thread():
+    signal_queue = queue.Queue()
+    thread = threading.Thread(target=_run_buffer, args=(signal_queue,))
+    thread.daemon = True
+    thread.start()
+    return signal_queue
+
+
+@pytest.mark.slowtest
+@requires_neuromag2ft
+def test_fieldtrip_rtepochs():
+    raw_tmax = 20
+    raw = mne.io.read_raw_fif(raw_fname, preload=True)
+    raw.crop(tmin=0, tmax=raw_tmax)
+    events_offline = mne.find_events(raw, stim_channel='STI 014')
+    event_id = list(np.unique(events_offline[:, 2]))
+    tmin, tmax = -0.2, 0.5
+    epochs_offline = mne.Epochs(raw, events_offline, event_id=event_id,
+                                tmin=tmin, tmax=tmax)
+    epochs_offline.drop_bad()
+    isi_max = (np.max(np.diff(epochs_offline.events[:, 0])) /
+               float(raw.info['sfreq'])) + 7.0
+
+    kill_signal = _start_buffer_thread()
+
+    try:
+        X_rt = None
+        events_ids_rt = None
+        with FieldTripClient(host='localhost', port=1972,
+                             tmax=raw_tmax, wait_max=2) as rt_client:
+            # get measurement info guessed by MNE-Python
+            raw_info = rt_client.get_measurement_info()
+            assert ([ch['ch_name'] for ch in raw_info['chs']] ==
+                    [ch['ch_name'] for ch in raw.info['chs']])
+
+            # create the real-time epochs object
+            rt_epochs = RtEpochs(rt_client, event_id, tmin, tmax,
+                                 stim_channel='STI 014', isi_max=isi_max)
+            rt_epochs.start()
+
+            time.sleep(0.5)
+            for ev_num, ev in enumerate(rt_epochs.iter_evoked()):
+                if ev_num == 0:
+                    X_rt = ev.data[None, :, :]
+                    events_ids_rt = int(
+                        ev.comment)  # comment attribute contains the event_id
+                else:
+                    X_rt = np.concatenate((X_rt, ev.data[None, :, :]),
+                                          axis=0)
+                    events_ids_rt = np.append(events_ids_rt, int(ev.comment))
+
+            rt_epochs.stop()
+
+        assert_array_equal(events_ids_rt, rt_epochs.events[:, 2])
+        assert_array_equal(X_rt, rt_epochs.get_data())
+        assert len(rt_epochs) == len(epochs_offline)
+        assert_array_equal(events_ids_rt, epochs_offline.events[:, 2])
+        # we don't get exactly the same values (type conversions?)
+        # check only that "almost all" values are close
```

Comment:
that's weird. Are you sure it's not a missing epoch?

Reply:
The numbers of epochs match. I inspected arrays in the debugger, some channels of the last epoch are incorrect. 

```
> percentage_of_correct_samples_per_ch = np.mean(np.isclose(rt_epochs.get_data()[-1, :, :],
                                                            epochs_offline.get_data()[-1, :, :]), 
                                                 axis=1)
> np.mean(percentage_of_correct_samples_per_ch[:314])
1.0
> np.mean(percentage_of_correct_samples_per_ch[314:])
0.03041912497126657
```
So somewhere the ``FieldTrip`` code returns invalid/incorrect buffer data (but not zeros)... Do you have a good prior where to look? It might as well be in the ``neuromag2ft`` code...

Reply:
okay that smells like a bug to me. I doubt it is in the `neuromag2ft` code though ... could it be some problem here: https://github.com/mne-tools/mne-python/blob/master/mne/realtime/fieldtrip_client.py#L350-L353 ?

Reply:
Nevermind, the bug was an off-by-one error, since ``externals.FieldTrip.Client.getData`` returns the data inclusive the ``stop`` index. I'm not sure though why this wasn't caught in ``RtEpochs`` due to the buffer being too short for event processing...